src/input_parsers/VnnLibParser.cpp

```diff
@@ -0,0 +1,549 @@
+ ** directory for licensing information.\endverbatim
+ **
+ ** A parser for properties encoded in VNN-LIB format.
+
+**/
+
+#include <regex>
+#include "File.h"
+#include "InputParserError.h"
+#include "VnnLibParser.h"
+#include "DisjunctionConstraint.h"
+
+static double extractScalar ( const String &token )
+{
+    std::string::size_type end;
+    double value = std::stod ( token.ascii (), &end );
+    if ( end != token.length ())
+    {
+        throw InputParserError ( InputParserError::UNEXPECTED_INPUT, Stringf ( "%s not a scalar",
+                                                                               token.ascii ()).ascii ());
+    }
+    return value;
+}
+
+static String readVnnlibFile ( const String &vnnlibFilePath )
+{
+    if ( !File::exists ( vnnlibFilePath ))
+    {
+        std::cout << "Error: the specified property file "
+                  << vnnlibFilePath.ascii () << " doesn't exist!" << std::endl;
+        throw InputParserError ( InputParserError::FILE_DOESNT_EXIST, vnnlibFilePath.ascii ());
+    }
+
+    File vnnlibFile ( vnnlibFilePath );
+    vnnlibFile.open ( File::MODE_READ );
+
+    String vnnlibContent;
+
+    try
+    {
+        while ( true )
+        {
+            String line = vnnlibFile.readLine ().trim ();
+            if ( line == "" || line.substring ( 0, 1 ) == ";" )
+            {
+                continue;
+            }
+            vnnlibContent += line;
+        }
+    }
+    catch ( const CommonError &e )
+    {
+        // A "READ_FAILED" is how we know we're out of lines
+        if ( e.getCode () != CommonError::READ_FAILED )
+            throw e;
+    }
+
+    return vnnlibContent;
+}
+
+void
+VnnLibParser::parse ( const String &vnnlibFilePath, InputQuery &inputQuery )
+{
+    String vnnlibContent = readVnnlibFile ( vnnlibFilePath );
+
+    parseVnnlib ( vnnlibContent, inputQuery );
+}
+
+void
+VnnLibParser::parseVnnlib ( const String &vnnlibContent, InputQuery &inputQuery )
+{
+    std::regex re ( R"(\(|\)|[\w\-\\.]+|<=|>=|\+|-|\*)" );
+
+    auto tokens_begin = std::cregex_token_iterator ( vnnlibContent.ascii (),
+                                                     vnnlibContent.ascii ()
+                                                     + vnnlibContent.length (), re );
+    auto tokens_end = std::cregex_token_iterator ();
+
+    Vector<String> all_tokens;
+
+    for ( std::cregex_token_iterator it = tokens_begin;
+          it != tokens_end; ++it )
+    {
+        auto match = *it;
+        auto match_str = String ( match.str ().c_str ());
+
+        all_tokens.append ( match_str );
+    }
+
+    parseScript ( all_tokens, inputQuery );
+}
+
+int
+VnnLibParser::parseScript ( const Vector<String> &tokens, InputQuery &inputQuery )
+{
+    int index = 0;
+
+    while (( unsigned ) index < tokens.size ())
+    {
+        ASSERT( tokens[index] == "(" )
+        index = parseCommand ( index + 1, tokens, inputQuery );
+        ASSERT( tokens[index] == ")" )
+
+        index++;
```

Comment:
always prefer ++index to index++ when both can be used; it is more efficient, generally